src/python/pants/backend/python/lint/isort/subsystem.py

```diff
@@ -65,6 +65,29 @@ class Isort(PythonToolBase):
             """
         ),
     )
+    firstparty_discovery = BoolOption(
+        default=False,
+        advanced=True,
+        help=softwrap(
+            """
+            If enabled, Pants will tell `isort` explicitly which imported modules are firstparty (as
+            decided by if Pants is aware of the module via `python_` targets).
+
+            Enabling this option works around a common "gotcha" of isort runs in Pants, where isort's
+            default heuristic of "its firstparty if the file exists locally" can lead to surprising
+            behavior in Pants where the sandbox for the run is only populated with a subset of the
+            files in repo, leading to isort marking firstparty modules as thirdparty.
```

Comment:
FWIW, I'd phrase this a bit more positively, especially if we're intending to flip it on by default: instead of it being a work-around, it's just how pants resolves the interaction between isort and sandboxing. E.g.

> When this option is enabled, Pants ensures that isort knows all modules that are provided by this repository, to include them in the `FIRSTPARTY` section. By default, isort's heuristic is "`FIRSTPARTY` if the file exists", but Pants runs multiple isort processes in parallel in sandboxes, and each sandbox only includes on a subset of the files in the repo, which can lead to isort considering some first-party modules as third-party.

Also, total nitpick, but the word in normal text is "first-party" or "first party", and I'd suggest using code'd `FIRSTPARTY` when referring to the all-one-word isort section configuration value. To make it even more confusing isort isn't very consistent: this is tying into the isort's `known_first_party` option... also separate words! I'd thus be slightly inclined to call the option `first_party_discovery`... but am not sure, `firstparty_discovery` is probably fine too 🤔